fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,147 @@
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        match self
+            .started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => return Err(anyhow!("already started")),
+        }
+
+        let config = self.config.clone();
+        let provider = self.parent_view_provider.clone();
+        let agent = self.agent.clone();
+
+        tokio::spawn(async move {
+            if let Err(e) = sync_with_parent(config, agent, provider).await {
+                tracing::info!("sync with parent encountered error: {e}");
+            }
+        });
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent<T: ParentFinalityProvider + Send + Sync + 'static>(
+    config: Config,
+    agent_proxy: Arc<IPCAgentProxy>,
+    provider: Arc<T>,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval));
+
+    loop {
+        interval.tick().await;
+
+        // update block hash and validator set
+        let latest_height = agent_proxy
+            .get_chain_head_height()
+            .await
+            .context("cannot fetch parent chain head")?;
+
+        if latest_height < config.chain_head_delay {
+            tracing::debug!("latest height not more than the chain head delay");
+            continue;
+        }
+
+        let starting_height = if let Some(h) = provider.latest_height().await {
+            h + 1
+        } else {
+            // if latest_recorded_height is None, it means we have not started fetching any heights
+            // we just use the latest height minus a lower bound as the parent view
+            max(
+                1,
+                latest_height.saturating_sub(config.chain_head_lower_bound),
+            )
+        };
+
+        for h in starting_height..=latest_height {
+            let block_hash = agent_proxy
+                .get_block_hash(h)
+                .await
+                .context("cannot fetch block hash")?;
+            let validator_set = agent_proxy
+                .get_validator_set(h)
+                .await
+                .context("cannot fetch validator set")?;
+
+            match provider
+                .new_parent_view(Some((h, block_hash, validator_set)), vec![])
+                .await
+            {
+                Ok(_) => {}
+                Err(e) => match e {
+                    Error::ParentReorgDetected(_) => {
+                        // FIXME: the most brutal way is to clear all the cache in `provider` and
+                        // FIXME: start from scratch.
+                        todo!()
+                    }
+                    _ => unreachable!(),
+                },
+            }
+        }
+
+        // now get the top down messages
+        let nonce = provider.latest_nonce().await.unwrap_or(0);
+        let top_down_msgs = agent_proxy.get_top_down_msgs(latest_height, nonce).await?;
```

Comment:
I see that this only depends on [what has been fetched](https://github.com/consensus-shipyard/fendermint/pull/205/files#diff-8edd9923a9eb262950fd919a858ece4449d8a41abf401bd3c1e4ac1d63190454R63-R64) in this session, not what has been committed before. 

Does that mean that after every restart we'll ask the agent to fetch all messages since genesis, with nonce = 0? I don't see how this could be what we want, how could all cross-messages fit into memory?

Reply:
right, it should be fetching from the last committed finality, similar to the block height. Not sure why I placed 0.